### ISSUE TITLE
Better error handling / railway oriented programming

### ISSUE THREAD

**Issue by BackDoorManUC** [dotnet/csharplang#2474]
C# is known for using exceptions everywhere - even in non-exceptional situations. 

Add error handling similar to Rust or F# because creating custom exceptions for everything is bad and there's no other standard way to deal with errors in C#.

- [Why Exceptions should be Exceptional](https://mattwarren.org/2016/12/20/Why-Exceptions-should-be-Exceptional/)
- [Railway Oriented Programming](https://fsharpforfunandprofit.com/rop/)

Libraries that help:

- https://github.com/mcintyre321/OneOf
- https://github.com/Resultful/Resultful
- https://github.com/gnaeus/OperationResult

Related issue: https://github.com/dotnet/csharplang/issues/113

**Comment by yaakov-h** [comment]
If there are libraries, then why do we need a language change?

What's the actual thing you want C# to do, that is currently not valid syntax?

**Comment by BackDoorManUC** [comment]
@yaakov-h 

> If there are libraries, then why do we need a language change?

to not need libraries - make it a part of BCL

> What's the actual thing you want C# to do, that is currently not valid syntax?

It's more https://github.com/dotnet/corefx/ issue but error handling is a language feature that's why I opened this here.

**Comment by stuartstein777** [comment]
Feel C# would need discriminated unions before you get railroad orientated programming and who knows if that will ever happen...

**Comment by DavidArno** [comment]
> The compiler should check if all errors are handled that will guarantee safety

That runs counter to the idea of railway orientated programming (ROP). It's not about checking if all errors are handled, it's about simply having two paths through the code, the happy path and the error path, switching away from the happy path to the error path on the first error and skipping all other happy path functionality.

The language can support this model better with eg support for DUs so that a first class either type can be supported by pattern matching etc, but that's not specific to ROP. Additionally, it could provide keywords around `bind` etc, but ROP is far too niche a concept at the moment to be incorporated into the language that much in my view.

**Comment by huoyaoyuan** [comment]
Exception is NOT SLOWER; it's faster when fault path is **not hit**. The deeper the call stack, the more exception saves. Language can help you getting rid of writing `if (success)`, but every implicit if does cost.
ROP can be considered as an **alternative**, not something better.

**Comment by BackDoorManUC** [comment]
@huoyaoyuan benchmarks prove exceptions are slower

**Comment by YairHalberstadt** [comment]
@BackDoorManUC
http://joeduffyblog.com/2016/02/07/the-error-model/ makes for a very interesting read in general. His tests showed that if exceptions are rare, they are actually faster on average than explicit error checking. However that was on Midoris version of exceptions, which were faster than C#s.

**Comment by BackDoorManUC** [comment]
@YairHalberstadt I mean exceptions shouldn't be used for regular control flow - for errors that are expected. It's one of the biggest flaws in C#. Fortunately, we have `Try*` method variants to avoid them.

https://stackoverflow.com/questions/729379/why-not-use-exceptions-as-regular-flow-of-control

Exceptions are fine, but unfortunately, they're abused because there's no proper error handling support.

**Comment by huoyaoyuan** [comment]
@BackDoorManUC This benchmark doesn't hit what I said. It does not investigate **deep TryXXX pattern**.

**Comment by BackDoorManUC** [comment]
@huoyaoyuan The blog post linked mentions TryXXX pattern. Could you elaborate?

**Comment by DavidArno** [comment]
The `TryXXX` pattern is achieved either through returning a `maybe` or returning a `boolean` and spitting out a value via an `out`. Either way, you either get a value or no value, rather than a value or exception.

**Comment by HaloFour** [comment]
Other than potentially supporting DUs (which is already championed) I don't see any language proposals coming out of this issue.  If you want result-types in the BCL, that would be a CoreFX issue.  I don't see that as being reasonable until C# figures out how DUs will work as to be useful it would have to fit that model, as well as be usable from all other languages in the ecosystem.

**Comment by DavidArno** [comment]
> … TryXXX pattern works in simple situations but it doesn't if you want more meaningful errors returned. If the method has more than 1 reason to fail

True, in that case, you use the "either" approach:
```cs
Either<Value, ErrorInfo> Xxxx() =>
```
or
```cs
bool Xxxx(out Value value, out ErrorInfo error) =>
```
In the latter case, `value` is set on a `true` being returned, else `error` is set. This approach is less common in C# due to the lack of built-in support for discriminated unions and because multiple `out's get messy very quickly.

**Comment by HaloFour** [comment]
Let me rephrase that:  Other than supporting DUs or something similar, are there any specific language proposals that you are requesting here?  What would that alternative error handling syntax look like?

I'll note that while I expect something like DUs to be able to fulfill the request here that I don't expect the use of exceptions for flow control to disappear or even to decrease is use appreciably.  For better or worse they're just too prevalent, and it's unlikely that the BCL will make non-throwing versions of all of the APIs that currently throw.  If DUs play really nicely with the rest of the .NET ecosystem I could see the BCL team perhaps adding a couple of "official" ones, particularly if they are interested in exposing some APIs using them, but otherwise they're more likely to be relegated to separate NuGet packages.

**Comment by orthoxerox** [comment]
Once I was trying to explain to myself why I found Rust's `?` so elegant. I thought, "Well, it lets you work with potentially fallible functions like they always succeed and handle errors higher up on the call stack where it's more convenient." Then I realized that's literally how exceptions work and started appreciating them more.

One case where exceptions fall short is regular, expected errors. There you might want to reuse the same "bail out early with minimal fuss" syntax that exceptions use, but C# doesn't support that. [ExceptionDispatchInfo](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.exceptionservices.exceptiondispatchinfo?view=netcore-2.2) might be a way out, but being able to throw anything would simplify this as well.

**Comment by asos-richardgibson** [comment]
As for language features that support this proposal, I'd have to say that F#'s computation expressions are almost certainly it. 

They're the feature that allows you to rewrite the following:

```fsharp
let x = 
	match getCustomer id with
	| None -> None
	| Some customer ->
		match customer.address with
		| None -> None
		| Some address ->
			match address.line1 with
			| None -> None
			| Some line -> line.Length 
```

As this (assuming you have written an` OptionBuilder` class and have an instance of it called `maybe`):

```fsharp
let x =
	maybe {
		let! customer = getCustomer id
		let! address = customer.address
		let! line = address.line1

		return line.Length
	}
```

It's a bit like having async / await available as a general-use language feature

**Comment by HaloFour** [comment]
@asos-richardgibson 

The first example definitely falls under DUs.  We're already getting `switch` expression support.

Although that's a really verbose way to write:  😁 

```cs
int? x = GetCustomer(id)?.Address?.Line1?.Length;
```

The second example seems to fall more under `for`/`do` comprehensions, which are very much like LINQ except for composing over general functor/monad type classes.  I could see that being useful to a limited degree but given the language already has LINQ and `await` the main use cases would seemingly already be covered.

**Comment by MgSam** [comment]
> @YairHalberstadt I mean exceptions shouldn't be used for regular control flow - for errors that are expected. It's one of the biggest flaws in C#. Fortunately, we have `Try*` method variants to avoid them.
> 
> https://stackoverflow.com/questions/729379/why-not-use-exceptions-as-regular-flow-of-control
> 
> > if you use exceptions for normal situations, how do you locate unusual (ie exceptional) situations
> 
> Exceptions are fine, but unfortunately, they're abused because there's no proper error handling support.

Well written methods do not use exceptions for control flow. You're right in that there are places in the BCL where exceptions are abused to convey non-exceptional information (File I/O coming to mind). However, this is entirely a library issue and not a language issue. There has been extensive debate in the CoreFx repo about improving those methods that use exceptions in this way (though I don't think the debates ever reached conclusions).

**Comment by Korporal** [comment]
> @huoyaoyuan The blog post linked mentions TryXXX pattern. Could you elaborate?

I find the TryXXX pattern very appealing and often leads to simplified more readable code. One issue though is the need for all these TryXXX methods and when you have a large codebase trying to convert it to leverage TryXXX more can be time consuming.

So something like this springs to mind, here we can pass any existing (non TryXXX) method or function into these generic wrappers and get the benefit right away of the pattern by simply wrapping our existing non TryXXX methods at their invocation.

```cs
using System;

namespace ConsoleApp9
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");

            if (TryWithoutExceptions(Console.ReadKey, out var keyinfo))
            {
                // Do stuff, we got the info...
            }

            if (TryWithExceptions<string, InvalidOperationException>(GetTextA, out var text))
            {
                // Do stuff, we got the info...
            }
        }

        static string GetTextA()
        {
            throw new InvalidOperationException();
        }

        /// <summary>
        /// Attempts to execute the supplied function and never throws an exception.
        /// </summary>
        /// <typeparam name="A">The type of the function's result.</typeparam>
        /// <param name="function">A function that returns a result.</param>
        /// <param name="result">Null if any kind of exception was thrown else a valid value.</param>
        /// <returns>True if the returned result is not null else false.</returns>
        static bool TryWithoutExceptions<A>(Func<A> function, out A result)
        {
            try
            {
                return TryWithExceptions<A, Exception>(function, out result);
            }
            catch (Exception)
            {
                result = default;
                return false;
            }
        }

        /// <summary>
        /// Attempts to execute the supplied function and will throw an exception of type E but mask all other exceptions.
        /// </summary>
        /// <typeparam name="A">The type of the function's result.</typeparam>
        /// <typeparam name="E">An exception type, if thrown by the function it will be rethrown to the caller.</typeparam>
        /// <param name="function">A function that returns a result.</param>
        /// <param name="result">Null if any kind of exception was thrown else a valid value.</param>
        /// <returns>True if the returned result is not null else false.</returns>
        static bool TryWithExceptions<A,E>(Func<A> function, out A result) where E : Exception
        {
            if (function == null)
                throw new ArgumentNullException(nameof(function));

            try
            {
                result = function();
                return true;
            }
            catch(E)
            {
                throw;
            }
            catch (Exception)
            {
                result = default;
                return false;
            }
        }
    }
}
```

I just put this together now but the attractive nature of TryXXX has been on my mind for some time and the above code reflects something I may do going forward.

The TryWithoutExceptions method guarantees you'll never get an exception and will always get true/false.

The TryWithExceptions method is the same but for the fact it will potentially throw if an exception of the type specified arises when calling the function.

Naturally there are variations on this simple model, including supporting more of the Func<> types.

**Comment by BackDoorManUC** [comment]
@Korporal TryXXX requires using `out` and this is not supported with `async` methods. I use Try methods when possible but sometimes it's just not enough. I hope to see more F# features at some point in the future. In F# you have `Result<T, TError>` https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/results

I think corefx repo will be more appropriate for this issue.
